pkg/proxy/proxy.go

```diff
@@ -0,0 +1,218 @@
+		return
+	}
+
+	p.serveReverseProxy(ctx, ns, res, req)
+}
+
+func responseWithError(res http.ResponseWriter, err *crterrors.Error) {
+	http.Error(res, err.Error(), err.Code)
+}
+
+// createContext creates a new gin.Context with the User ID extracted from the Bearer token.
+// To be used for storing the user ID and logging only.
+func (p *Proxy) createContext(req *http.Request) (*gin.Context, error) {
+	userID, err := p.extractUserID(req)
+	if err != nil {
+		return nil, err
+	}
+	keys := make(map[string]interface{})
+	keys[context.SubKey] = userID
+	return &gin.Context{
+		Keys: keys,
+	}, nil
+}
+
+func (p *Proxy) getTargetNamespace(ctx *gin.Context) (*namespace.Namespace, error) {
+	userID := ctx.GetString(context.SubKey)
+	return p.namespaces.GetNamespace(ctx, userID)
+}
+
+func (p *Proxy) extractUserID(req *http.Request) (string, error) {
+	userToken, err := extractUserToken(req)
+	if err != nil {
+		return "", err
+	}
+
+	token, err := p.tokenParser.FromString(userToken)
+	if err != nil {
+		return "", crterrors.NewUnauthorizedError("unable to extract userID from token", err.Error())
+	}
+	return token.Subject, nil
+}
+
+func extractUserToken(req *http.Request) (string, error) {
+	a := req.Header.Get("Authorization")
+	token := strings.Split(a, "Bearer ")
+	if len(token) < 2 {
+		return "", crterrors.NewUnauthorizedError("no token found", "a Bearer token is expected")
+	}
+	return token[1], nil
+}
+
+func (p *Proxy) serveReverseProxy(ctx *gin.Context, target *namespace.Namespace, res http.ResponseWriter, req *http.Request) {
+	proxy := p.newReverseProxy(ctx, target)
+
+	// Note that ServeHttp is non blocking and uses a go routine under the hood
+	proxy.ServeHTTP(res, req)
+}
+
+func (p *Proxy) newReverseProxy(ctx *gin.Context, target *namespace.Namespace) *httputil.ReverseProxy {
+	targetQuery := target.APIURL.RawQuery
+	director := func(req *http.Request) {
+		origin := req.URL.String()
+		req.URL.Scheme = target.APIURL.Scheme
+		req.URL.Host = target.APIURL.Host
+		req.URL.Path = singleJoiningSlash(target.APIURL.Path, req.URL.Path)
+		log.Info(ctx, fmt.Sprintf("forwarding %s to %s", origin, req.URL.String()))
+		if targetQuery == "" || req.URL.RawQuery == "" {
+			req.URL.RawQuery = targetQuery + req.URL.RawQuery
+		} else {
+			req.URL.RawQuery = targetQuery + "&" + req.URL.RawQuery
+		}
+		if _, ok := req.Header["User-Agent"]; !ok {
+			// explicitly disable User-Agent so it's not set to default value
+			req.Header.Set("User-Agent", "")
+		}
+		// Replace token
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", target.TargetClusterToken))
+	}
+	var transport *http.Transport
+	if !configuration.GetRegistrationServiceConfig().IsProdEnvironment() {
+		transport = &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+	}
+	return &httputil.ReverseProxy{
+		Director:      director,
+		Transport:     transport,
+		FlushInterval: -1,
+	}
+}
+
+func singleJoiningSlash(a, b string) string {
```

Comment:
this would deserve a few unit tests but I wonder if we could use `filepath.Join` instead

Reply:
I would avoid using `filepath.Join` because it's OC specific.
Added more tests in https://github.com/codeready-toolchain/registration-service/pull/224/commits/d633a95af25cbc705ba48f0ce35afcd03c05bdaa